monai/transforms/compose.py

```diff
@@ -148,26 +72,22 @@ def execute_compose(
         start: the index of the first transform to be executed. If not set, this defaults to 0
         end: the index after the last transform to be exectued. If set, the transform at index-1
             is the last transform that is executed. If this is not set, it defaults to len(transforms)
-        lazy_evaluation: whether to enable lazy evaluation for lazy transforms. If False, transforms will be
+        lazy: whether to enable lazy evaluation for lazy transforms. If False, transforms will be
             carried out on a transform by transform basis. If True, all lazy transforms will
             be executed by accumulating changes and resampling as few times as possible.
             A `monai.transforms.Identity[D]` transform in the pipeline will trigger the evaluation of
             the pending operations and make the primary data up-to-date.
         overrides: this optional parameter allows you to specify a dictionary of parameters that should be overridden
             when executing a pipeline. These each parameter that is compatible with a given transform is then applied
-            to that transform before it is executed. Note that overrides are currently only applied when lazy_evaluation
-            is True. If lazy_evaluation is False they are ignored.
+            to that transform before it is executed. Note that overrides are currently only applied when lazy
+            is True. If lazy is False they are ignored.
             currently supported args are:
             {``"mode"``, ``"padding_mode"``, ``"dtype"``, ``"align_corners"``, ``"resample_mode"``, ``device``},
             please see also :py:func:`monai.transforms.lazy.apply_transforms` for more details.
         override_keys: this optional parameter specifies the keys to which ``overrides`` are to be applied. If
             ``overrides`` is set, ``override_keys`` must also be set.
         threading: whether executing is happening in a threaded environment. If set, copies are made
             of transforms that have the ``RandomizedTrait`` interface.
-        log_stats: whether to log the detailed information of data and applied transform when error happened,
-            for NumPy array and PyTorch Tensor, log the data shape and value range,
-            for other metadata, log the values directly. default to `False`.
-        verbose: whether to print debugging info when lazy_evaluation=True.
```

Comment:
why remove these two flags?

Reply:
They didn't match with the rest of the API. @ericspod and I were discussing it and he suggested that these parameters should be replaced with 'logger_name', which allows you to override the default logger, as this is how it is done in the rest of the code, then everything goes to the logger with either level 'info' or 'error', depending on what bits are being logged. Looks like I have forgotten to put it in so I need to do it still

Reply:
@wyli are you fine with the `logger_name`? It is consistent with other parts of the API

Reply:
currently `verbose=True` prints pending operation counts:
```
2023-04-26 10:59:48,011 - INFO - Lazy - 'image' - upcoming: 'EnsureTyped'- pending 0
2023-04-26 10:59:48,012 - INFO - Lazy - 'label' - upcoming: 'EnsureTyped'- pending 0
2023-04-26 10:59:48,709 - INFO - Lazy - 'image' - upcoming: 'EnsureChannelFirstd'- pending 0
2023-04-26 10:59:48,709 - INFO - Lazy - 'label' - upcoming: 'EnsureChannelFirstd'- pending 0
2023-04-26 10:59:48,709 - INFO - Lazy - 'image' - upcoming: 'Orientationd'- pending 0
2023-04-26 10:59:48,709 - INFO - Lazy - 'label' - upcoming: 'Orientationd'- pending 0
2023-04-26 10:59:48,712 - INFO - Lazy - 'image' - upcoming: 'Spacingd'- pending 1
2023-04-26 10:59:48,712 - INFO - Lazy - 'label' - upcoming: 'Spacingd'- pending 1
2023-04-26 10:59:48,721 - INFO - Lazy - 'image' - upcoming: 'CastToTyped'- pending 2
2023-04-26 10:59:48,721 - INFO - Lazy - 'label' - upcoming: 'CastToTyped'- pending 2
2023-04-26 10:59:48,722 - INFO - Lazy - 'image' - upcoming: 'CropForegroundd'- pending 2
2023-04-26 10:59:48,722 - INFO - Lazy - 'label' - upcoming: 'CropForegroundd'- pending 2
```
it's useful for debugging the transforms, what's the equivalent usage in this PR?

Reply:
I can put something roughly equivalent in.

Reply:
@wyli logging of pending is back in

Reply:
thanks, it's almost there, I think we need logs for each key such as `image` and `label`  to debug them separately...also now the `_apply_transform` in transforms/transform.py need some refactoring to decouple the concepts, 

 The current dev branch has a much cleaner version https://github.com/Project-MONAI/MONAI/blob/4f291723c4ce1a1baf7efbd499495b777dfb1869/monai/transforms/transform.py#L64-L67

Reply:
It already handles dictionary based transforms by key. The test doesn't reflect that but is a quick addendum.

You can't really separate the various different methods:
On dev we have:
`evaluate_with_overrides`, `apply_transforms`, `apply_transform` and `_apply_transform`
On this branch we have:
`apply_pending_transforms`, `apply_pending`, `apply_transform` and `_apply_transform`

In both cases, these four methods are all quite interdependent (they rely on each other for certain pre and post conditions to be the case). I've spread the functionality across them differently so that none of them has overwhelming complexity, and they do more on the branch than on dev (for example the lambda problem is fixed and we have the enabled mode (lazy = None on compose) that will execute transforms according the the flags they have set on them, allowing for more fine tuning).

Reply:
I specifically moved a bit of that responsibility into _apply_transforms as it reduced the need to have a separate call-stack doing the evaluate_with_overrides bit. That is where so much of the complexity in that function comes from. Moving the functionality there simplified everything hugely while also allowing for a more general solution

Reply:
Disagree...I think the apply_transform logic should be improved, also using logger name to trigger logging looks like a workaround, what if in the future we want different levels of verbosity? The core elements must be simple and readable

Reply:
That is a different matter. It would be a good idea to have a discussion about logging in general across the app and determine what meets the needs of MONAI for all its scenarios. I followed Eric's advice about making it at least consistent with other places that use logging. I'd target a reworked logging mechanism as a 1.3 task

Reply:
`evaluate_with_overrides, apply_transforms, apply_transform and _apply_transform`

 is far more complex than what I've replaced it with. It also fixed issues like the issues that you were having with Lambda and it introduced a third lazy mode.

`apply_pending_transforms, apply_pending, apply_transform and _apply_transform`

Reply:
What is the lambda issue? You mentioned multiple times and don't actually get it. Could you please file a ticket about it?

Reply:
The issue where you end up with a Lambda function in the wrong place in the applied operations sequence (it should be at the end but it ended up at the start because it got applied while a tensor had pending operations (the lazy integration test). It doesn't need a ticket because you guys are aware of it and have done several PRs to modify code and tests to account for it

Reply:
FYI - dictionary based transform output
```
            "Accumulate pending transforms - lazy mode: True, key: 'a', pending: 0, upcoming 'Flipd', transform.lazy: False (overridden)\n"
            "Accumulate pending transforms - lazy mode: True, key: 'b', pending: 0, upcoming 'Flipd', transform.lazy: False (overridden)\n"
Accumulate pending transforms - lazy mode: True, key: 'a', pending: 1, upcoming 'Spacingd', transform.lazy: False (overridden)
Accumulate pending transforms - lazy mode: True, key: 'b', pending: 1, upcoming 'Spacingd', transform.lazy: False (overridden)
Accumulate pending transforms - lazy mode: True, key: 'a', pending: 2, upcoming 'Rotate90d', transform.lazy: False (overridden)
Accumulate pending transforms - lazy mode: True, key: 'b', pending: 2, upcoming 'Rotate90d', transform.lazy: False (overridden)
Apply pending transforms - lazy mode: True, key: 'a', pending: 3, upcoming 'NormalizeIntensityd', transform is not lazy
```

Reply:
This PR description says it resolves an issue with lambda, so it would be great to report the issue first. it's unclear to me what's not working and what's the new behaviour after this fix.

Reply:
tests\test_integration_lazy_samples.py
The pipeline has a LambdaD at the end of it. That LambdaD gets put out of place in the applied transforms, which blows up the invert. Hence the code that you added at https://github.com/Project-MONAI/MONAI/blob/4f291723c4ce1a1baf7efbd499495b777dfb1869/tests/test_integration_lazy_samples.py#L152

Reply:
This isn't needed because the LambdaD can't get out of place in the applied operations with the modified lazy resampling execution

Reply:
#6439

Reply:
I think we spend enough time on this PR and try to merge it before v1.2 release, still don't agree on a few points discussed here, and it's also risky to merge this fairly large PR as we are already in codefreeze for v1.2. Let's revisit this after the release. For now I don't have time to discuss/improve it further.

by the way the logging still doesn't work, to reproduce:

```
import torch
import monai.transforms as mt

data = torch.zeros((1, 40, 40))
xform = mt.Compose([mt.Flip(0), mt.Rotate90(1), mt.Zoom(0.8)], lazy=True, logger_name="log")
xform(data)

```

doesn't output anything.

Reply:
Probably this is an issue of logging configuration. The logging tests as passing.

This is ready to go (pending :
. It fixes issues that exist (or existed) on dev
. It passes all tests
. It is more comprehensively documented

Reply:
Hi All,

This PR has several important features for 1.2, as it is critical for several stuff we are working on. Can we please try to merge this before 1.2 is released? I still see many PRs with new features being merged, so I don't quite get what the issue is.

Reply:
> @mjorgecardoso  I still see many PRs with new features being merged, so I don't quite get what the issue is.
> 

we are currently at the 2d quality assurance drop, 5th release candidate of v1.2, we shouldn't merge new feature like this into the release candidate. please could you help me understand 'many PRs with new features'?

Reply:
@atbenmurray logging -- expected behaviour:
![Screenshot 2023-04-27 at 13 35 46](https://user-images.githubusercontent.com/831580/234863667-5502f376-800d-48dd-add3-118c6b619bb3.png)

this branch removes this feature, and in my opinion, replaced it with a worse design ...

Reply:
@wyli I am taking a look at this

Reply:
New commit should resolve that. The only issue was that I needed to use `monai.app.utils.get_logger` to ensure that the named logger was properly configured

Reply:
```
2023-04-27 17:56:05,617 - INFO - Accumulate pending transforms - lazy: True, pending: 0, upcoming 'Flip', transform.lazy: False (overridden)
2023-04-27 17:56:05,622 - INFO - Accumulate pending transforms - lazy: True, pending: 1, upcoming 'Rotate90', transform.lazy: False (overridden)
2023-04-27 17:56:05,625 - INFO - Accumulate pending transforms - lazy: True, pending: 2, upcoming 'Zoom', transform.lazy: False (overridden)
torch.Size([1, 40, 40, 40])
```

Reply:
ok sure, marking this as resolved...

Reply:
Additional modification

```
import torch
import monai.transforms as mt

data = torch.zeros((1, 40, 40, 40))
xform = mt.Compose([mt.Flip(0), mt.Rotate90(1), mt.Zoom(0.8), mt.ApplyPending(), mt.Spacing(1.2)],
                   lazy=True, logger_name="log")
xform(data)
```

gives

```
2023-04-27 18:31:14,550 - INFO - Accumulate pending transforms - lazy: True, pending: 0, upcoming 'Flip', transform.lazy: False (overridden)
2023-04-27 18:31:14,552 - INFO - Accumulate pending transforms - lazy: True, pending: 1, upcoming 'Rotate90', transform.lazy: False (overridden)
2023-04-27 18:31:14,554 - INFO - Accumulate pending transforms - lazy: True, pending: 2, upcoming 'Zoom', transform.lazy: False (overridden)
2023-04-27 18:31:14,557 - INFO - Apply pending transforms - lazy: True, pending: 3, upcoming 'ApplyPending', transform is not lazy
2023-04-27 18:31:14,567 - INFO - Accumulate pending transforms - lazy: True, pending: 0, upcoming 'Spacing', transform.lazy: False (overridden)
2023-04-27 18:31:14,575 - INFO - Pending transforms applied: applied_operations: 4
```

Reply:
I think this should be sufficient:
```
import torch
import monai.transforms as mt

data = torch.zeros((1, 40, 40, 40))
xform = mt.Compose([mt.Flip(0), mt.Rotate90(1), mt.Zoom(0.8), mt.ApplyPending(), mt.Spacing(1.2)],
                   lazy=True, logger_name="log")
xform(data)
```

gives

```
2023-04-27 18:35:32,310 - INFO - Accumulate pending transforms - lazy: True, pending: 0, upcoming 'Flip', transform.lazy: False (overridden)
2023-04-27 18:35:32,313 - INFO - Accumulate pending transforms - lazy: True, pending: 1, upcoming 'Rotate90', transform.lazy: False (overridden)
2023-04-27 18:35:32,314 - INFO - Accumulate pending transforms - lazy: True, pending: 2, upcoming 'Zoom', transform.lazy: False (overridden)
2023-04-27 18:35:32,317 - INFO - Apply pending transforms - lazy: True, pending: 3, upcoming 'ApplyPending', transform is not lazy
2023-04-27 18:35:32,326 - INFO - Pending transforms applied: applied_operations: 3
2023-04-27 18:35:32,327 - INFO - Accumulate pending transforms - lazy: True, pending: 0, upcoming 'Spacing', transform.lazy: False (overridden)
2023-04-27 18:35:32,335 - INFO - Pending transforms applied: applied_operations: 4
```

Reply:
I need to modify the logging test to remove the timestamp. This will probably go in first thing tomorrow morning

Reply:
I ran the example using this PR, got:
```
Traceback (most recent call last):
  File "my_test.py", line 5, in <module>
    xform = mt.Compose([mt.Flip(0), mt.Rotate90(1), mt.Zoom(0.8), mt.ApplyPending(), mt.Spacing(1.2)],
AttributeError: module 'monai.transforms' has no attribute 'ApplyPending'

```

could you confirm @atbenmurray ?

Reply:
this is resolved..